solidity/random-beacon/contracts/DKGValidator.sol

```diff
@@ -189,6 +195,42 @@ contract DKGValidator {
             actualGroupMembersHash;
     }
 
+    /// @notice Performs validation of active group members as declared in DKG
+    ///         result against all members and misbehaved members. This function
+    ///         can be used only after all the members and misbehaved members
+    ///         validity has been checked.
+    /// @return true if active group members are valid. False otherwise.
+    function validateActiveGroupMembers(DKG.Result calldata result)
+        internal
+        view
+        returns (bool)
+    {
+        // If no misbehaved members, then 'members' array should be equal to
+        // 'activeMembers' array
+        if (result.misbehavedMembersIndices.length == 0) {
+            return
+                keccak256(abi.encode(result.members)) ==
+                keccak256(abi.encode(result.activeMembers));
+        }
+
+        for (uint256 i = 0; i < result.misbehavedMembersIndices.length; i++) {
+            // group member indices start from 1, so we need to -1 on misbehaved
+            uint256 misbehavedMemberArrayPosition = result
+                .misbehavedMembersIndices[i - 1] - 1;
+            for (uint256 j = 0; j < result.activeMembers.length; j++) {
+                // active members array should not contain any misbehaved members
+                if (
+                    result.members[misbehavedMemberArrayPosition] ==
+                    result.activeMembers[j]
+                ) {
+                    return false;
+                }
```

Comment:
I do not think this validation rule is correct. One operator could have multiple seats in a group. It is possible that one member from that operator misbehaved and another did not. Then, we should expect only one of them to be removed from the group. That's how it worked in v1 and that's how off-chain DKG is implemented.

Reply:
Ok, for simplicity say the group size is 10. We have:
`misbehavedMembersIndices = [1,7,3]`
`members = [2,4,1,5,2,8,7,9,5,4]`

`members[1] = 4`
`members[7] = 9`
`members[3] = 5`

These ^ are the ids of misbehaved members. Ex:
- `activeMembers = [2,3,1,2,8,7,9,6,8]` -> invalid active members, `9` misbehaved
- `activeMembers = [2,3,1,2,8,7,4,6,8]` -> invalid active members, `4` misbehaved
- `activeMembers = [2,3,1,4,8,7,4,6,8]` -> invalid active members. `4` id is present twice, but it is the same member id that misbehaved.
- `activeMembers = [2,3,1,6,8,7,6,8]` -> valid active members, no misbehaved

In this function we do not remove anything, it returns `true/false` if the `activeMembers` array `valid/invalid. Can you pls point where's a flaw here, so I can understand it better using this example?

Reply:
I see two problems with the current implementation.

First one:
```
for (uint256 i = 0; i < result.misbehavedMembersIndices.length; i++) {
  // group member indices start from 1, so we need to -1 on misbehaved
  uint256 misbehavedMemberArrayPosition = result
    .misbehavedMembersIndices[i - 1] - 1;
```

What is `result.misbehavedMembersIndices[i - 1]` for `i = 0`?

I am surprised tests are green because Solidity documentation says:

> By default, arithmetic is always “checked”, which mean that if the result of an operation falls outside the value range of the type, the call is reverted through a failing assertion. 

https://docs.soliditylang.org/en/v0.8.10/types.html#integers

Similar example is in [this article](https://blog.soliditylang.org/2020/10/28/solidity-0.8.x-preview/) in *Actual Checked Arithmetic* section.

It is possible we do not have it properly covered in unit tests but I'd be surprised if that's the case given that it should fail on the simplest case with even just one misbehaved member. 🤔 



Second problem:

The current implementation is going through `misbehavedMemberIndices`, turns them into operator identifiers, and checks if these operator identifiers are not present in `activeMembers`.

This is not how `setGroupMembers` worked and it is not how DKG protocol is designed and implemented.

Using your example:

```
members = [2,4,1,5,2,8,7,9,5,4]
misbehavedMembersIndices = [1,7,3]
```
so:
```
 1 2 3 4 5 6 7 8 9 10
[2,4,1,5,2,8,7,9,5,4]
 x   x       x
```

Note that `misbehavedMemberIndices` are 1-indexed, just like member IDs and that's why we need to `-1`, see [here](https://github.com/keep-network/keep-core/blob/main/solidity/random-beacon/contracts/libraries/DKG.sol#L65-L68).

Active members (without changing the order) should be: `[4,5,2,8,9,5,4]`.

@nkuba to double-check

Reply:
Just to note, this potential change is meant to be implemented in a separate PR which was agreed on Discord and in some [comments](https://github.com/keep-network/keep-core/pull/2768#discussion_r773103929). In this discussion, I just wanted to make sure if we want to explore this further and if there are some major flaws.

As for the first problem, there are some background findings by me after I opened this PR. One of them is test coverage of this verification step. I've changed a bunch of tests in this PR but overlooked covering this particular function. It will fail just like you said, and `misbehavedMembersIndices[i - 1] - 1;` should be probably replaced with `misbehavedMembersIndices[i] - 1;` (TBD in a separate PR).

As for the second problem, I've skipped `-1` for simplicity, just wanted to discuss a general idea. Implementation includes that. From your example, I see that what I was missing is if `activeMembers` is trimmed/replaced with some valid members like `[4,5,2,8,9]` (`[5,4]` are missing), then the verification would still pass, but it shouldn't. The `activeMembers` should be exactly the subset of `members` minus those that misbehaved, not more and not less, correct?

Reply:
Let me rephrase: we are marking group members with the particular indexes as misbehaving. We are not marking operators with the particular ids as misbehaving. That's the difference. 

In other words, it is possible that the operator 0xABC with identifier (from sortition pool) 1337 could be selected to the group at three positions: 1, 50, 64, and it is misbehaving at positions 1 and 50. It means that we should remove operator 1337 at positions 1 and 50 but leave them at position 64.

Reply:
Right, I think I understood you. 
```
members = [2,4,1,5,2,8,7,9,5,4]
operator 0xABC -> sortition pool id 1337 was selected to the group at positions [1,7,3]
but this operator is misbehaving at [1,7] 
```
marking [1,7] as misbehaved but leaving an index [3]
```
 1 2 3 4 5 6 7 8 9 10
[2,4,1,5,2,8,7,9,5,4]
 x           x
```
the active members array should be exactly:
```
[4,1,5,2,8,9,5,4]
```

Reply:
Please also note the specific way how `setGroupMembers` removes entries by replacing them with poped from the end of the list.
So in the example above it would result with:
```
[4,4,1,5,2,8,5,9]
```

Reply:
Yes, I noticed, but I thought we can pass members that participated in dkg in order that we specify, no? I know that the `hash([4,1,5,2,8,9,5,4]) != hash([4,4,1,5,2,8,5,9])`, but can't we filter all members off-chain by just removing misbehaved, so the original order is preserved? Additional shuffling of members adds complexity on-chain, which most likely adds gas. Unless I'm not aware of something and it has to be in the order you've mentioned.